Added Unix Implementation for omrsock_getaddrinfo and omrsock_freeaddrinfo in OMRSOCK API

caohaley opened the pull request:
- Added New Error Code for getaddrinfo related failures
- Added Unix implementations for OMRSOCK getaddrinfo and freeaddrinfo
- Added new tests relating getaddrinfo and freeaddrinfo

Issue: #4794

caohaley commented:
@babsingh  @rwy0717 This is ready for review I believe. The first 6 commits are from the element extraction PR. The last 4 commits is for this PR.

caohaley commented:
@babsingh @rwy0717 I updated this PR with the merge of the element extraction PR this morning. Now ready for review. There are 4 commits in total for this PR.

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,32 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
-{
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+{	
+	if ((NULL == hints) || (NULL == result)) {
```
It should be safe for hints to be null, in which case we need to pass a null hints object to the OS call.

babsingh commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -237,7 +237,40 @@ TEST(PortSockTest, create_hints_and_element_extraction)
+	/* Testing invalid arguments: getaddrinfo pointer to result is NULL. */
+	omrsock_addrinfo_t resultPtr = NULL;
+
+	rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, (char *)"localhost", NULL, hints, resultPtr);
```
Is `(char *)` needed?
```suggestion
	rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, "localhost", NULL, hints, resultPtr);
```

caohaley replied:
@babsingh I remembered why I added it in, it seems like the conversation from string to char* is deprecated for some reason. Should I add it back in?
```/root/omr/fvtest/porttest/omrsockTest.cpp:261:83: warning: deprecated conversion from string constant to 'char*' [-Wwrite-strings]
  rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, "localhost", NULL, hints, resultPtr);
                                                                                   ^
/root/omr/fvtest/porttest/omrsockTest.cpp:269:81: warning: deprecated conversion from string constant to 'char*' [-Wwrite-strings]
  rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, "localhost", NULL, hints, &result);
                                                                                 ^
/root/omr/fvtest/porttest/omrsockTest.cpp:291:80: warning: deprecated conversion from string constant to 'char*' [-Wwrite-strings]
  rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, "localhost", NULL, NULL, &result);
```

babsingh commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -237,7 +237,40 @@ TEST(PortSockTest, create_hints_and_element_extraction)
+	rc = OMRPORTLIB->sock_freeaddrinfo(OMRPORTLIB, resultPtr);
+	EXPECT_EQ(rc, OMRPORT_ERROR_INVALID_ARGUMENTS);
+
+	/* Get and verify that omrsock_getaddrinfo and omrsock_freeaddrinfo works.  */
```
```suggestion
	/* Get and verify that omrsock_getaddrinfo and omrsock_freeaddrinfo works. */
```

babsingh commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -237,7 +237,40 @@ TEST(PortSockTest, create_hints_and_element_extraction)
+	EXPECT_EQ(rc, OMRPORT_ERROR_INVALID_ARGUMENTS);
+
+	/* Get and verify that omrsock_getaddrinfo and omrsock_freeaddrinfo works.  */
+	rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, (char *)"localhost", NULL, hints, &result);
```
```suggestion
	rc = OMRPORTLIB->sock_getaddrinfo(OMRPORTLIB, "localhost", NULL, hints, &result);
```

babsingh commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -237,7 +237,40 @@ TEST(PortSockTest, create_hints_and_element_extraction)
+	EXPECT_EQ(rc, OMRPORT_ERROR_INVALID_ARGUMENTS);
+
+	/* Testing invalid arguments: freeaddrinfo pointer to result is NULL. */
+
```
```suggestion
```

babsingh commented on fvtest/porttest/omrsockTest.cpp:
```diff
@@ -237,7 +237,40 @@ TEST(PortSockTest, create_hints_and_element_extraction)
+	EXPECT_NE(length, 0);
+
+	OMRPORTLIB->sock_freeaddrinfo(OMRPORTLIB, &result);
+
```
```suggestion
```

babsingh commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,32 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+	
+	omr_os_addrinfo *addrInfoResults = NULL;
+	omr_os_addrinfo *addrInfoHints = NULL;
+	int32_t count = 0;
```
Variable declaration should always be placed at the start of the function before the `if` statement. Windows compilers don't allow it. If this code is used for the Windows implementation, then it won't work. You should always write code which is compiler agnostic.

babsingh commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,32 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+	
+	addrInfoHints = (omr_os_addrinfo *)hints->addrInfo;
+
+	/* Verify that getaddrinfo was successful */
```
```suggestion
	/* Verify that getaddrinfo was successful. */
```

babsingh commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,32 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+		count++;
+		addrInfoResults = addrInfoResults->ai_next;
+	}
+	/* We didn't count the first entry */
```
avoid first person.
```suggestion

	/* Account for the first entry. */
```

babsingh commented:
WIP can be removed.

caohaley commented:
@babsingh @rwy0717 Ready for another round of review.

caohaley commented:
@babsingh @rwy0717 I added one more test case to omrsockTest.cpp. I will leave it like this now and wait for your reviews. Thank you!

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,40 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+
+	addrInfoHints = (omr_os_addrinfo *)hints->addrInfo;
+
+	/* Verify that getaddrinfo was successful. */
```
```suggestion
```

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,40 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+	if (0 != getaddrinfo(node, service, addrInfoHints, &addrInfoResults)) {
+		return OMRPORT_ERROR_SOCK_ADDRINFO_FAILED;
+	}
+	memset(result, 0, sizeof(struct OMRAddrInfoNode));
```
If the result pointer is valid, I think we should clear it, even (especially?) if there was an error. I think think memset should move to line 88 above.

caohaley replied:
Yes, that makes sense, I will change that.

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -157,7 +189,16 @@ omrsock_getaddrinfo_protocol(struct OMRPortLibrary *portLibrary, omrsock_addrinf
+	freeaddrinfo((omr_os_addrinfo *)handle->addrInfo);
+
+	handle->addrInfo = NULL;
+	handle->length = 0;		
```
```suggestion
	handle->length = 0;
```

babsingh commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,40 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+	}
+	
+	if (NULL == hints) {
+		rc = portLibrary->sock_getaddrinfo_create_hints(portLibrary, &hints, 0, 0, 0, 0);
```
who will free (remove) the hints?

rwy0717 replied:
This is incorrect behaviour, the right thing to do is:
```c++
omr_os_addrinfo* os_hints = NULL;
if (NULL != hints) {
  os_hints = hints->addrInfo;
}
```

[original impl](https://github.com/eclipse/openj9/blob/master/runtime/port/unix/j9sock.c#L856-L858)

caohaley replied:
@rwy0717 @babsingh Actually, I used `omr_os_addrinfo *addrInfoHints = NULL;` instead of `os_hints` but the functionality, I think should be the same because in the next line is 
```addrInfoHints = (omr_os_addrinfo *)hints->addrInfo;```.
I was thinking that if hints is NULL, then we will create a hints for them using all 0s which means just no requirement on family, socket type, protocol or flags. This means like if it doesn't fail, then hints after this will no longer be NULL and an invalid argument. If it does fail, we can return the error from the function call (in omrsock_getaddrinfo_create_hints).

The hints allocation is using the per thread buffer so I think it'll replace current values in the hints buffer with all 0s. I was thinking after we set the &hints (using omrsock_getaddrinfo_create_hints) so that hints point to the per thread buffer with all 0s, we can use the new hints to pass into the getaddrinfo system function later on.

caohaley commented:
@rwy0717 @babsingh Ready for another round of review.

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,35 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+		addrInfoHints = (omr_os_addrinfo *)hints->addrInfo;
+	}
+
+	if (0 != getaddrinfo(node, service, addrInfoHints, &addrInfoResults)) {
```
Is it possible for getaddrinfo to return 0, but still set result to NULL (ie no matching results)?

caohaley replied:
I believe it has to be 1 or more addrinfo node to be successful:
"Upon successful return, a pointer to a linked list of one or more addrinfo structures is returned through the res parameter."
"Given node and service, which identify an Internet host and a service, getaddrinfo() returns one or more addrinfo structures, each of which contains an Internet address that can be specified in a call to bind(2) or connect(2)."

rwy0717 replied:
Makes sense, thanks 👍

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,35 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
+		addrInfoResults = addrInfoResults->ai_next;
+	}
+	/* Account the first entry */
+	result->length = count++;
```
I would initialize count to 1 before the while loop, rather than adding 1 at the end, it makes more sense to me.

rwy0717 commented on port/unix/omrsock.c:
```diff
@@ -76,8 +76,35 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
 omrsock_getaddrinfo(struct OMRPortLibrary *portLibrary, char *node, char *service, omrsock_addrinfo_t hints, omrsock_addrinfo_t result)
-{
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+{	
```
```suggestion
{
```

rwy0717 commented:
@genie-omr build all

caohaley commented:
@rwy0717  Sorry, I found out that there was a warning when I removed the cast from string to (char *), I added it back in just now, could you run it again?

rwy0717 commented:
Actually, it would be better if we changed the signature of omrsock_getaddrinfo, so that the node and service arguments are const, rather than casting away the constness.

Too bad we don't have an omr_socka

caohaley commented:
Oh I seee, so it is because C++ char array is constant but C it isn't constant. Should I change it or should I leave it with a cast?

And no worries! Hahaha

caohaley commented:
There seems to be another error where I set i to be int32_t and comparing that to length, which is unsigned int. I will change it but I'm going to wait for another 5 minutes to see if it picks up any other errors too.

caohaley commented:
The int32_t and uint32_t conversation is a little tricky, is it alright if I cast the length to int32_t in the test?

I can't change i to be unsigned int because the index is signed int, unless I change index to unsigned int. To change the index, I may need to change a few files and need to open a new PR to do that.

```	OMRPORTLIB->sock_getaddrinfo_length(OMRPORTLIB, &result, &length);
	ASSERT_NE(length, 0);

	for (int32_t i = 0; i < (int32_t)length; i++) {
		rc = OMRPORTLIB->sock_getaddrinfo_family(OMRPORTLIB, &result, &family, i);
		EXPECT_EQ(rc, 0);
		EXPECT_EQ(family, hintsFamily);

		rc = OMRPORTLIB->sock_getaddrinfo_socktype(OMRPORTLIB, &result, &sockType, i);
		EXPECT_EQ(rc, 0);
		EXPECT_EQ(sockType, hintsSockType);

		rc = OMRPORTLIB->sock_getaddrinfo_protocol(OMRPORTLIB, &result, &protocol, i);
		EXPECT_EQ(rc, 0);
	}
```

youngar commented:
@genie-omr build all

youngar commented on include_core/omrport.h:
```diff
@@ -2114,11 +2114,11 @@ typedef struct OMRPortLibrary {
 	int32_t (*sock_getaddrinfo_length)(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, uint32_t *length) ;
 	/** see @ref omrsock.c::omrsock_getaddrinfo_family "omrsock_getaddrinfo_family"*/
-	int32_t (*sock_getaddrinfo_family)(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *family, int32_t index )	;
+	int32_t (*sock_getaddrinfo_family)(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *family, uint32_t index )	;
```
Nitpick: There is a `\t` before the `;` when it should be a space.  Why there is a space beats me ¯\_(ツ)_/¯

caohaley replied:
Yes, I've always wondered too! But I just followed the convention XD

youngar commented on port/unix/omrsock.c:
```diff
@@ -134,14 +163,14 @@ omrsock_getaddrinfo_socktype(struct OMRPortLibrary *portLibrary, omrsock_addrinf
-omrsock_getaddrinfo_protocol(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *protocol, int32_t index)
+omrsock_getaddrinfo_protocol(struct OMRPortLibrary *portLibrary, omrsock_addrinfo_t handle, int32_t *protocol, uint32_t index)
 {
 	if ((NULL == handle) || (NULL == handle->addrInfo) || (index < 0) || (index >= handle->length)) {
```
From the compiler warning, index is never less than 0.
```suggestion
	if ((NULL == handle) || (NULL == handle->addrInfo) || (index >= handle->length)) {
```

caohaley replied:
Sounds good, they have been fixed.

youngar commented:
@genie-omr build osx

youngar commented:
@genie-omr build all

youngar commented:
@babsingh Will you do a final review?

youngar commented:
@genie-omr build all

babsingh commented on port/unix/omrsock.c:
```diff
@@ -77,7 +77,36 @@ omrsock_getaddrinfo_create_hints(struct OMRPortLibrary *portLibrary, omrsock_add
-	return OMRPORT_ERROR_NOT_SUPPORTED_ON_THIS_PLATFORM;
+	omr_os_addrinfo *addrInfoResults = NULL;
+	omr_os_addrinfo *addrInfoHints = NULL;
+	int32_t count = 0;
```
`OMRAddrInfoNode->length` is `uint32_t` and `count` will never be negative. so, `count` should be `uint32_t`.

youngar commented:
@genie-omr build all

youngar commented:
@genie-omr build all

youngar commented:
@genie-omr build all
